### PR TITLE
lib/systems: strip kernel to avoid reference cycles

### LINKED ISSUE #224694: Build failure: linux (powerpc)
Building powerpc kernels (I only tested pkgsCross.powernv, but examination of the kernel source code suggests to me that 32-bit powerpc, and well as big-endian 64-bit will be the same) is broken since https://github.com/NixOS/nixpkgs/pull/221707.  This happens because on powerpc, the kernel target we install is vmlinux, rather than any derivative of it like zImage.  The kernel build system will strip when creating such a derivative image, but because we install vmlinux directly, we get an unstripped image.  This means the kernel contains references to its build directory, which since that PR will be the $dev output.

AIUI, 64-bit POWER boots via kexec, and a zImage is a self-extracting kernel ELF image, so hopefully the fix would just be to switch the kernel target for powerpc to zImage, and things will work again, and POWER users will get smaller kernel images as a bonus.

I'm not sure about 32-bit PowerPC, but this [kernel documentation](https://www.kernel.org/doc/html/latest/powerpc/bootwrapper.html) might be helpful.

If we can't use an *Image target for any powerpc targets, there is also a target called vmlinux.strip that we could fall back to, but let's take the opportunity to evaluate whether anything else is more suitable first, when we're changing this stuff anyway.

Apart from powerpc, the only [target](https://github.com/NixOS/nixpkgs/blob/cd1e297b9327ff3242227264977e4bea2b747d81/lib/systems/platforms.nix#L407-L483) we have that uses vmlinux as its kernel target is a MIPS machine called the Fuloong 2.0.  I don't know how it boots.

### LINKED ISSUE #244438: kernel builds on mips broken by #221707

### PR DESCRIPTION
**Pull request by ghost**
###### Description of changes

lib/systems: strip kernel to avoid reference cycles

Since https://github.com/NixOS/nixpkgs/pull/221707 we must ensure    that the image resulting from a kernel build is stripped.  If we do    not do this, there will be circular dependencies among its outpaths.

This commit fixes the breakage on mips.

Closes #244438

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by vkleen** [comment]
It turns out that I can't actually get the PowerPC zImage to boot with kexec. Sorry for the half-baked suggestion... Unfortunately, for reasons that I don't understand, doing the same stripping on `vmlinux` as you're doing for MIPS doesn't seem to be enough to get rid of cyclic references.

**Comment by amjoseph-nixpkgs** [comment]
> It turns out that I can't actually get the PowerPC zImage to boot with kexec. Sorry for the half-baked suggestion... Unfortunately, for reasons that I don't understand, doing the same stripping on `vmlinux` as you're doing for MIPS doesn't seem to be enough to get rid of cyclic references.

MIPS kernels are actually fully ELF-compliant images, so a lot of stuff is easier there.

File: pkgs/os-specific/linux/kernel/manual-config.nix [hunk 1]
```diff
@@ -376,6 +379,10 @@ stdenv.mkDerivation ({
   preFixup = ''
     # Don't strip $dev/lib/modules/*/vmlinux
     stripDebugList="$(cd $dev && echo lib/modules/*/build/*/)"
+  '' + lib.optionalString (stdenv.hostPlatform.isPower64) ''
+    $STRIP -s $out/zImage || true
+  '' + lib.optionalString (stdenv.hostPlatform.isMips) ''
+    $STRIP -s $out/vmlinux || true
```
Comment (alyssais) [inline pkgs/os-specific/linux/kernel/manual-config.nix:385]: This is what I was trying to avoid — I'd rather have the stripping done by the kernel build system where it supports it, because they'll do a better job of setting the right flags, I think.

On MIPS, there's the `vmlinux.bin` target, which is stripped.  Does that work?

File: lib/systems/platforms.nix [hunk 2]
```diff
@@ -28,7 +28,10 @@ rec {
 
       baseConfig = "powernv_defconfig";
       target = "vmlinux";
```
Comment (alyssais) [inline lib/systems/platforms.nix:30]: Isn't this the thing to change?

**Review by alyssais — DISMISSED** [review]

File: lib/systems/platforms.nix (hunk 2 above)
Comment (vkleen) [reply lib/systems/platforms.nix:30]: This needs to be changed as well but doesn't actually affect the kernel build system. The `target` field is only used by the bootloader install scripts in NixOS, as far as I can tell.

Comment (alyssais) [reply lib/systems/platforms.nix:30]: https://github.com/NixOS/nixpkgs/blob/78f5472f4a792679879ac6a8cc0806dd1d9cac59/pkgs/os-specific/linux/kernel/manual-config.nix#L299-L305

File: pkgs/os-specific/linux/kernel/manual-config.nix (hunk 1 above)
Comment (amjoseph-nixpkgs) [reply pkgs/os-specific/linux/kernel/manual-config.nix:385]: > I'd rather have the stripping done by the kernel build system where it supports it

Okay, if that's your goal I will see what I can do.

> because they'll do a better job of setting the right flags, I think.

Well, I'm not sure I agree with that.  I've been dismayed at how little of the build process machinery is shared across architectures.  In a lot of places they just reimplement the same thing multiple times, and then those implementations drift out of sync with each other.  MIPS is especially bad here.

But I will give it a try.

File: lib/systems/platforms.nix (hunk 2 above)
Comment (amjoseph-nixpkgs) [reply lib/systems/platforms.nix:30]: That convoluted chunk of code won't actually result in `kernelConf="zImage"` causing the build process to invoke `make zImage` :frown:

Comment (amjoseph-nixpkgs) [reply lib/systems/platforms.nix]: > Isn't this the thing to change?

Implemented in 9298d32eab6bf636a2aa594a725effc8c0195faa

Testing now.

File: pkgs/os-specific/linux/kernel/manual-config.nix (hunk 1 above)
Comment (amjoseph-nixpkgs) [reply pkgs/os-specific/linux/kernel/manual-config.nix]: Implemented in latest push.  Currently testing.

File: pkgs/os-specific/linux/kernel/manual-config.nix [hunk 3]
```diff
@@ -230,6 +230,8 @@ stdenv.mkDerivation ({
     # replicated here to apply to older versions.
     # Makes __FILE__ relative to the build directory.
     "KCPPFLAGS=-fmacro-prefix-map=$(sourceRoot)/="
+  ] ++ lib.optionals (kernelConf?target) [
+    "KBUILD_IMAGE=${kernelConf.target}"
```
Comment (alyssais) [inline pkgs/os-specific/linux/kernel/manual-config.nix:234]: Ah yes, great.  I've been wanting to do this for a while, but haven't had a reason because (at least for aarch64) it doesn't work for the really old kernels.  For POWER/MIPS, maybe that's fine.

File: lib/systems/default.nix [hunk 4]
```diff
@@ -119,7 +119,14 @@ rec {
...
         ({
-          linux-kernel = args.linux-kernel or {};
+          # The following per-cpu defaults `target` ensure that the
+          # image is stripped, which is now necessary in order to
+          # avoid circular dependencies.
+          linux-kernel = lib.optionalAttrs final.isMips {
+            target = "vmlinux.bin";
+          } // lib.optionalAttrs final.isPower {
+            target = "vmlinux.strip.gz";
+          } // (args.linux-kernel or {});
```
Comment (alyssais) [inline lib/systems/default.nix:129]: Can you explain why this change is necessary? Doesn't platforms.nix already take care of it?  (Although I realise we're missing a big-endian POWER definition in platforms.nix.)

Comment (amjoseph-nixpkgs) [reply lib/systems/default.nix:129]: It's in the commit message (copied below).  The upshot is that `platforms.nix` and `examples.nix` don't have "least upper bounds" -- there may be an assortment of `lib.systems.examples` entries which use CPU XYZ, but there isn't always (or even usually) some root `lib.systems.examples.xyz` which all of them extend from.

```
These defaults are placed within `lib.systems.default` rather than
`lib.systems.examples` since there isn't always a clear inheritance
hierarchy in the latter, let alone a "least upper bound of all
machines with CPU X".  The only other possible place where this
default could be placed would be in `pkgs/os-specific/linux/kernel`,
but at the moment that file isn't structured in a way that makes it
easy to achieve all three of the following goals:

1. Set a default value based on CPU type
2. Ensure that the default will be overridden by more-specific
   choices in `lib/systems/examples.nix`
3. Make it unlikely that refactorings of
   `pkgs/os-specific/linux/kernel` will accidentally lose the
   defaults.
```

Comment (alyssais) [reply lib/systems/default.nix:129]: I had read the commit message but still didn't understand.



> there isn't always (or even usually) some root `lib.systems.examples.xyz` which all of them extend from.

Isn't this what the platforms.nix definitions are?

File: pkgs/os-specific/linux/kernel/manual-config.nix (hunk 1 above)
Comment (amjoseph-nixpkgs) [reply pkgs/os-specific/linux/kernel/manual-config.nix]: > On MIPS, there's the `vmlinux.bin` target, which is stripped. Does that work?

No, it doesn't.

File: lib/systems/default.nix (hunk 4 above)
Comment (amjoseph-nixpkgs) [reply lib/systems/default.nix:129]: No.  Maybe they should be, but they certainly aren't.

File: pkgs/os-specific/linux/kernel/manual-config.nix (hunk 3 above)
Comment (amjoseph-nixpkgs) [reply pkgs/os-specific/linux/kernel/manual-config.nix:234]: It doesn't work there either.  I can't get this to work and I've run out of time.

**Comment by amjoseph-nixpkgs** [comment]
@alyssais [wrote](https://github.com/NixOS/nixpkgs/pull/228018#pullrequestreview-1406908571)

> > This is what I was trying to avoid — I'd rather have the stripping done by the kernel build system where it supports it, because they'll do a better job of setting the right flags, I think.

I know that, stylistically, you were trying to avoid doing things this way, but no other solutions for the breakage caused by https://github.com/NixOS/nixpkgs/pull/221707 have arisen.  I can't get anything else to work and I have run out of time.

**Comment by alyssais** [comment]
(Originally posted [here](https://github.com/NixOS/nixpkgs/issues/224694#issuecomment-1559919474) by mistake.)

I've done some investigation of my own.  Here are the results:

## PowerPC

PowerPC ends up with two sets of references.  One is from kexec purgatory, and is probably an oversight.  I expect I could upstream a patch to fix this, like I recently did for x86 and RISC-V[^1].  The other is from the VDSO.  Unlike every other architecture, PowerPC deliberately uses a non-stripped vdso, apparently because it made perf traces nicer.  So unless things I've changed there (I'll try to have a look — it seems a bit strange that PowerPC would be special here), we won't be able to get the kernel build system to give us a reference-free PowerPC image.  So even though I don't like it, ack to stripping manually on PowerPC.

[^1]: the only thing getting in the way here is that it's a bit annoying to test, because our kexec test is [currently broken](https://github.com/NixOS/nixpkgs/pull/207039#issuecomment-1555289200).  cc @RaitoBezarius.

## MIPS

MIPS offers vmlinux, vmlinuz, and uImage.  vmlinuz and uImage should both be reference-free, but I don't know enough about how MIPS boots to know whether either of those would be sensible default targets.  I'd appreciate your insight here.  If it's not feasible to change to either of those images, then we probably have to strip manually on MIPS too.

**Comment by amjoseph-nixpkgs** [comment]
> MIPS offers vmlinux, vmlinuz, and uImage. vmlinuz and uImage should both be reference-free, but I don't know enough about how MIPS boots to know whether either of those would be sensible default targets.

`make vmlinuz` requires `CONFIG_SYS_SUPPORTS_ZBOOT`, which isn't supported by all mips machines.  Specifically Cavium Octeon routers, the most popular mips64 machines, don't have `CONFIG_SYS_SUPPORTS_ZBOOT` (the vendor-supplied bootloader is simply awful).

So `make vmlinuz` gives you:

```
CONFIG_SYS_SUPPORTS_ZBOOT is not enabled
```

I tried `make uImage` but it appears that is the same as `make vmlinux`, plus perhaps additional steps afterwards.  It's also uboot-specific?

> So even though I don't like it, ack to stripping manually on PowerPC.

@alyssais would you please update your review to reflect this?  It still shows the big-red-changes-requested-of-doom.

**Comment by amjoseph-nixpkgs** [comment]
Rebased.

**Comment by amjoseph-nixpkgs** [comment]
@ofborg build pkgsCross.powernv.linux

**Comment by amjoseph-nixpkgs** [comment]
From #194153, which includes this PR:

> pkgsCross.octeon.linux on x86_64-linux — Success [Details](https://github.com/NixOS/nixpkgs/pull/194153/checks?check_run_id=15183221426)

**Comment by amjoseph-nixpkgs** [comment]
I have moved the powerpc fix into

https://github.com/NixOS/nixpkgs/pull/244436

**Comment by amjoseph-nixpkgs** [comment]
Rebased.

It seems that ofborg-eval is on vacation today.
